crates/accelerate/src/quantum_circuit/circuit_data.rs

```
@@ -251,10 +251,13 @@ impl CircuitData {
                 "The number of qubits in the circuit has exceeded the maximum capacity",
             )
         })?;
-        self.qubit_indices_native.insert(BitAsKey::new(bit)?, idx);
-        self.qubits_native.push(bit.into_py(py));
-        self.qubits = PyList::new(py, &self.qubits_native).into_py(py);
-        Ok(())
+        self.qubit_indices_native
+            .try_insert(BitAsKey::new(bit)?, idx)
+            .map_or(Ok(()), |_| {
+                self.qubits_native.push(bit.into_py(py));
+                self.qubits = PyList::new(py, &self.qubits_native).into_py(py);
+                Ok(())
+            })
```

Do we definitely want to return `Ok(())` if the bit is already in the circuit data?  Feels like that should be a Python-space exception - we can do that with no cost to the happy path.  (Same comment for clbits.)

Either way, we might be stretching the legibility of the chained mapping methods - the current form might be clearer as
```rust
if self.qubit_indices_native.try_insert(...).is_ok() {
    self.qubits_native.push(bit.into_py(py));
    self.qubits = ...
}
Ok(())
```

---

Improved readability in ce055ca06a69fe05189234fef77a28608c099d77.

As it is, `ControlFlowBlockBuilder.append` relies on this behavior to avoid checking if an instruction's bits are already added. If we make this an error, we'll either need to maintain parallel sets for our bit lists in the builder, or catch and swallow the would-be exception.

I don't have a strong preference.

---

I think in most cases Qiskit's pretty vocal about forbidding adding bits that already exist to collections.  Do you think it would be worth making this method `strict=True` by default (raising an exception) and having the builder constructs use `strict=False`?

I'm fine with either direction: whichever you think is better for general use.

---

I like the `strict` kw-only option. Done in 62a724c7e45b9caa86f3eddf521cdea790859355.